services/pull/patch.go

```diff
@@ -98,139 +96,187 @@ func TestPatch(pr *models.PullRequest) error {
-	if err != nil {
-		return false, err
-	}
-	prConfig := prUnit.PullRequestsConfig()
+	// 5b. Create a plain patch from head to base
+	go func() {
+		if err := git.NewCommand(args...).RunInDirPipeline(tmpBasePath, diffWriter, nil); err != nil {
+			log.Error("Unable to get patch file from %s to %s in %s Error: %v", pr.MergeBase, pr.HeadBranch, pr.BaseRepo.FullName(), err)
+			_ = diffWriter.CloseWithError(err)
+			return
+		}
+		_ = diffWriter.Close()
+	}()
 
-	// 5. Prepare the arguments to apply the patch against the index
-	args := []string{"apply", "--check", "--cached"}
+	args = args[:0]
+
+	// 5c. Prepare the arguments to apply the patch against the index
+	args = append(args, "apply", "--check", "--cached")
 	if prConfig.IgnoreWhitespaceConflicts {
 		args = append(args, "--ignore-whitespace")
 	}
-	args = append(args, patchPath)
+	args = append(args, "-")
 	pr.ConflictedFiles = make([]string, 0, 5)
 
-	// 6. Prep the pipe:
+	// 5d. Prep the pipe:
 	//   - Here we could do the equivalent of:
 	//  `git apply --check --cached patch_file > conflicts`
 	//     Then iterate through the conflicts. However, that means storing all the conflicts
 	//     in memory - which is very wasteful.
 	//   - alternatively we can do the equivalent of:
 	//  `git apply --check ... | grep ...`
 	//     meaning we don't store all of the conflicts unnecessarily.
-	stderrReader, stderrWriter, err := os.Pipe()
-	if err != nil {
-		log.Error("Unable to open stderr pipe: %v", err)
-		return false, fmt.Errorf("Unable to open stderr pipe: %v", err)
-	}
+	stderrReader, stderrWriter := io.Pipe()
 	defer func() {
 		_ = stderrReader.Close()
 		_ = stderrWriter.Close()
 	}()
 
-	// 7. Run the check command
+	// 5e. Run the apply --check command
 	conflict := false
-	err = git.NewCommand(args...).
-		RunInDirTimeoutEnvFullPipelineFunc(
-			nil, -1, tmpBasePath,
-			nil, stderrWriter, nil,
-			func(ctx context.Context, cancel context.CancelFunc) error {
-				// Close the writer end of the pipe to begin processing
-				_ = stderrWriter.Close()
-				defer func() {
-					// Close the reader on return to terminate the git command if necessary
-					_ = stderrReader.Close()
-				}()
-
-				const prefix = "error: patch failed:"
-				const errorPrefix = "error: "
-
-				conflictMap := map[string]bool{}
-
-				// Now scan the output from the command
-				scanner := bufio.NewScanner(stderrReader)
-				for scanner.Scan() {
-					line := scanner.Text()
-					if strings.HasPrefix(line, prefix) {
-						conflict = true
-						filepath := strings.TrimSpace(strings.Split(line[len(prefix):], ":")[0])
+	go func() {
+		if err := git.NewCommand(args...).RunInDirTimeoutEnvFullPipeline(nil, -1, tmpBasePath, nil, stderrWriter, diffReader); err != nil {
+			_ = stderrWriter.CloseWithError(err)
+			return
+		}
+		_ = stderrWriter.Close()
+	}()
+	const empty = "error: unrecognised input"
+
+	const prefix = "error: patch failed:"
+	const errorPrefix = "error: "
+
+	conflictMap := map[string]bool{}
```

Comment:
why using map?

Reply:
So unfortunately the same file can be listed multiple times.